spec/abci++/abci++_basic_concepts_002_draft.md

```diff
@@ -1,51 +1,266 @@
+Tendermint gathers outstanding transactions from the mempool, generates a block header, and uses
+them to create a block to propose. Then, it calls `RequestPrepareProposal`
+with the newly created proposal, called _raw block_. The Application can
+make changes to the raw proposal, such as modifying transactions, and returns
+the (potentially) modified proposal, called _prepared proposal_ in the
+`Response*` call. The logic modifying the raw block can be non-deterministic.
+
+* [**ProcessProposal:**](./abci++_methods_002_draft.md#processproposal) It allows a validator to perform application-dependent work in a proposed block. This enables features such as allowing validators to reject a block according to whether the state machine deems it valid, and changing the block execution pipeline. Tendermint calls it when it receives a proposal. The Application cannot
+modify the proposal at this point but can reject it if it realizes it is invalid.
+If that is the case, Tendermint will prevote `nil` on the proposal, which has
+strong liveness implications for Tendermint. As a general rule, the Application
+SHOULD accept a prepared proposal passed via `ProcessProposal`, even if a part of
+the proposal is invalid (e.g., an invalid transaction); the Application can
+ignore the invalid part of the prepared proposal at block execution time.
+
+* [**ExtendVote:**](./abci++_methods_002_draft.md#extendvote) It allows applications to force their validators to do more than just validate within consensus. `ExtendVote` allows applications to include non-deterministic data, opaque to Tendermint, to precommit messages (the final round of voting). The data, called _vote extension_, will be part of the proposed block in the next height, along with the vote it is extending.
+The vote extension data is split into two parts, one signed by Tendermint as part
+of the vote data structure, and the other (optionally) signed by the Application.
+The Application may also choose not to include any vote extension.
+Tendermint calls it when is about to send a non-`nil` precommit message. 
+
+* [**VerifyVoteExtension:**](./abci++_methods_002_draft.md#verifyvoteextension) It allows validators to validate the vote extension data attached to a precommit message. If the validation fails, the precommit message will be deemed invalid and ignored
+by Tendermint. This has a negative impact on Tendermint's liveness, i.e., if repeatedly vote extensions by correct validators cannot be verified by correct validators, Tendermint may not be able to finalize a block even if sufficiently many (+2/3) of the validators send precommit votes for that block. Thus, `VerifyVoteExtension` should be used with special care.
+As a general rule, an Application that detects an invalid vote extension SHOULD
+accept it in `ResponseVerifyVoteExtension` and ignore it in its own logic. Tendermint calls it when 
+a process receives a precommit message with a (possibly empty) vote extension.
```

Comment:
Shall I open an issue about this?

Reply:
It seems like it's worth addressing. If there _is_ a use case for which we need to accept the risk to liveness, I guess we should at least document it.

Reply:
This is indeed a core point of ABCI++. Extensive discussions on this happened in December. I wrote the minutes of those discussions [here](https://github.com/tendermint/spec/pull/366#issuecomment-988669447) and [here](https://github.com/tendermint/spec/pull/366#issuecomment-991007915)... and maybe some other place I've just missed.
The second link above is the final meeting where we collectively decided to take the approach described in this text. You can check the slides, and we also recorded the meeting (link to recording is there). I remember it was a very interesting (and heated :-D) discussion we had.
Unless there are new concerns about this approach, I'd like us to stand by the decision we took back then.

Reply:
In case you don't want to go through the whole discussion again...
IIRC, the main reason that motivated our decision to have Tendermint reject the precommit message if `VerifyVoteExtension` returns _reject_ was that there are compelling use cases (threshold encryption of TXs) that needed to reject those messages in order to work, even if that might have an impact in Tendermint's liveness.
Dev should have more details on those use cases.

Reply:
It sounds like this has already been covered. I don't think it's a great compromise, but if that's what was agreed I don't have any new data to add.